### ISSUE TITLE
Build breaks if `pagePath` contains `%` character

### ISSUE THREAD

**Issue by Leksat** [gatsbyjs/gatsby#36350]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

As it is stated in [Gatsby docs](https://www.gatsbyjs.com/docs/reference/routing/creating-routes/), page paths must not be pre-encoded.

If a page path contains `%` character, the build breaks.

https://github.com/Leksat/my-gatsby-site/blob/5e54e7cf6aa2e4fac914d5c6a6b4c00d25af5e33/gatsby-node.ts#L7

<img width="1286" alt="Screen Shot 2022-08-09 at 16 36 53" src="https://user-images.githubusercontent.com/989015/183857738-b7a3c5bd-5ee0-4176-8ca7-c8770be6abc0.png">

This is because `decodeURIComponent` is applied to paths in `@gatsbyjs/reach-router`:
https://github.com/gatsbyjs/reach-router/blob/ccbdf3e10d6501093abd4f15e80f6261e5c7b036/src/lib/utils.js#L68

### Reproduction Link

https://github.com/Leksat/my-gatsby-site/blob/5e54e7cf6aa2e4fac914d5c6a6b4c00d25af5e33/gatsby-node.ts#L7

### Steps to Reproduce

1. Create a page with `/page/%` path
2. Run `gatsby build`

### Expected Result

The build passes

### Actual Result

The build fails with `URIError: URI malformed`

### Environment

```shell
System:
    OS: macOS 12.5
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 14.19.0 - ~/.nvm/versions/node/v14.19.0/bin/node
    Yarn: 1.22.17 - ~/.nvm/versions/node/v14.19.0/bin/yarn
    npm: 6.14.16 - ~/.nvm/versions/node/v14.19.0/bin/npm
  npmPackages:
    gatsby: ^4.20.0 => 4.20.0
```

**Comment by Leksat** [comment]
At the moment I have no idea how to fix this 😬

**Comment by j0sh77** [comment]
Thanks for filing this (and https://github.com/gatsbyjs/gatsby/issues/36348). We'll look into the best steps forward.

**Comment by Auspicus** [comment]
@Leksat I recently added those lines to the docs and wondered myself if Gatsby should support % and other special characters in the path. What is your take here? See: https://github.com/gatsbyjs/gatsby/issues/16765#issuecomment-1203476158 and https://github.com/gatsbyjs/gatsby/pull/36325

**Comment by Leksat** [comment]
Gatsby already supports almost every possible character in the path. So I'd try to fix this issue.

I already tried to add some tests to Gatsby source, but... Gatsby is a huge project, so I failed 😅 

In my company we use unicode/special-char paths (they come from a CMS) and we publish our Gatsby sites on Netlify without any issue. However, currently [we remove some characters/patterns](https://github.com/AmazeeLabs/silverback-mono/blob/16d683cd42111ce35e67fc12f70ca12798b040fb/packages/composer/amazeelabs/silverback_gatsby/src/Plugin/GraphQL/SchemaExtension/SilverbackGatsbySchemaExtension.php#L507-L526) from the paths for safety reasons based on my recommendations in https://github.com/gatsbyjs/gatsby/discussions/36345#discussioncomment-3364844

Whether we fix this particular issue or not, there still can be buggy paths, like `/my/../../page`. So maybe `createPage` should throw warnings if it sees unsafe paths?
